doc/neps/ufunc-overrides.rst

```diff
@@ -153,49 +152,207 @@ tuple in the ``out`` keyword argument.
+Generally, it is desirable to mirror the class hierarchy in the ufunc
+type casting hierarchy. The recommendation is that an
+``__array_ufunc__`` implementation of a class should generally return
+`NotImplemented` unless the inputs are instances of the same class or
+superclasses.  This guarantees that in the type casting hierarchy,
+superclasses are below, subclasses above, and other classes are
+incompatible (sadly, the terminology for graphs and classes has reversed
+vertical sense).  Exceptions to this need to check they respect the
+implicit type casting hierarchy.
+
+Subclasses can be easily constructed if methods consistently use
+:func:`super` to pass through the class hierarchy [7]_.  To support
+this, :class:`ndarray` has its own ``__array_ufunc__`` method,
+equivalent to::
+
+    def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
+        # Cannot handle items that have __array_ufunc__ (other than our own).
+        outputs = kwargs.get('out', ())
+        for item in inputs + outputs):
+            if (hasattr(item, '__array_ufunc__') and
+                    type(item).__array_ufunc__ is not ndarray.__array_ufunc__):
+                return NotImplemented
+
+        # If we didn't have to support legacy behaviour (__array_prepare__,
+        # __array_wrap__, etc.), we might here convert python floats,
+        # lists, etc, to arrays with
+        # items = [np.asarray(item) for item in inputs]
+        # and then start the right iterator for the given method.
+        # However, we do have to support legacy, so call back into the ufunc.
+        # Its arguments are now guaranteed not to have __array_ufunc__
+        # overrides, and it will do the coercion to array for us.
+        return getattr(ufunc, method)(*items, **kwargs)
+
+Note that, as a special case, the ufunc dispatch mechanism does not call
+this `ndarray.__array_ufunc__` method, even for `ndarray` subclasses
+if they have not overridden the default `ndarray` implementation. As a
+consequence, calling `ndarray.__array_ufunc__` will not result to a
+nested ufunc dispatch cycle.
+
+The use of :func:`super` should be particularly useful for subclasses of
+:class:`ndarray` that only add an attribute like a unit.  In their
+`__array_ufunc__` implementation, such classes can do possible
+adjustment of the arguments relevant to their own class, and pass on to
+the superclass implementation using :func:`super` until the ufunc is
+actually done, and then do possible adjustments of the outputs.
+
+In general, custom implementations of `__array_ufunc__` should avoid
+nested dispatch cycles.  However, for some subclasses, it may be better
+to use ``getattr(ufunc, method)(*items, **kwargs)``. For instance, for a
+class like :class:`MaskedArray`, which only cares that whatever
+it contains is an :class:`ndarray` subclass, a reimplementation with
+``__array_ufunc__`` is probably more easily done by directly applying
+the ufunc to its data, and then adjusting the mask.
```

Comment:
Or, dispatch to `self.value.__array_ufunc__`? --- There's a distinction in what types are handled, in the sense that a nested ufunc call will raise TypeErrors --- so I think if you call the ufunc, you need to first check that you handle the types of the inputs. Whereas, passing it on to `self.value.__array_ufunc__` delegates the responsibility of type checking.